src/lightning_fabric/loggers/tensorboard.py

```diff
@@ -0,0 +1,310 @@
+from lightning_fabric.utilities.logger import _add_prefix, _convert_params, _flatten_dict
+from lightning_fabric.utilities.logger import _sanitize_params as _utils_sanitize_params
+from lightning_fabric.utilities.rank_zero import rank_zero_only, rank_zero_warn
+from lightning_fabric.utilities.types import _PATH
+
+log = logging.getLogger(__name__)
+
+_TENSORBOARD_AVAILABLE = RequirementCache("tensorboard")
+_TENSORBOARDX_AVAILABLE = RequirementCache("tensorboardX")
+if TYPE_CHECKING:
+    # assumes at least one will be installed when type checking
+    if _TENSORBOARD_AVAILABLE:
+        from torch.utils.tensorboard import SummaryWriter
+    else:
+        from tensorboardX import SummaryWriter  # type: ignore[no-redef]
+
+
+class TensorBoardLogger(Logger):
+    r"""
+    Log to local file system in `TensorBoard <https://www.tensorflow.org/tensorboard>`_ format.
+
+    Implemented using :class:`~tensorboardX.SummaryWriter`. Logs are saved to
+    ``os.path.join(root_dir, name, version)``. This is the recommended logger in Lightning Fabric.
+
+    Args:
+        root_dir: The root directory in which all your experiments with different names and versions will be stored.
+        name: Experiment name. Defaults to ``'lightning_logs'``. If it is the empty string then no per-experiment
+            subdirectory is used.
+        version: Experiment version. If version is not specified the logger inspects the save
+            directory for existing versions, then automatically assigns the next available version.
+            If it is a string then it is used as the run-specific subdirectory name,
+            otherwise ``'version_${version}'`` is used.
+        default_hp_metric: Enables a placeholder metric with key `hp_metric` when `log_hyperparams` is
+            called without a metric (otherwise calls to ``log_hyperparams`` without a metric are ignored).
+        prefix: A string to put at the beginning of all metric keys.
+        sub_dir: Sub-directory to group TensorBoard logs. If a ``sub_dir`` argument is passed
+            then logs are saved in ``/root_dir/name/version/sub_dir/``. Defaults to ``None`` in which case
+            logs are saved in ``/root_dir/name/version/``.
+        \**kwargs: Additional arguments used by :class:`tensorboardX.SummaryWriter` can be passed as keyword
+            arguments in this logger. To automatically flush to disk, `max_queue` sets the size
+            of the queue for pending logs before flushing. `flush_secs` determines how many seconds
+            elapses before flushing.
+
+
+    Example::
+
+        from lightning.fabric.loggers import TensorBoardLogger
+
+        logger = TensorBoardLogger("path/to/logs/rot", name="my_model")
+        logger.log_hyperparams({"epochs": 5, "optimizer": "Adam"})
+        logger.log_metrics({"acc": 0.75})
+        logger.finalize("success")
+    """
+    LOGGER_JOIN_CHAR = "-"
+
+    def __init__(
+        self,
+        root_dir: _PATH,
```

Comment:
since we don't have the default root dir from trainer here, should we combine `root_dir`, `name` and `sub_dir` to a single path?

Reply:
I'm not sure we want to deviate that strongly from the design yet. I believe it's too early to tell. Eventually this logger interface will again be used by the Trainer through Lite, and we don't know whether there will be a "default_root_dir" or not. 

Btw, subdir gets concatenated after the version. A full path would look like this

`root_dir/name/version/sub_dir`

not

`root_dir/name/sub_dir/version`

Reply:
I know, but IMO we should simplify as much as possible here. If I see a logger I expect there to be an option with one path and I'd have it log exactly at that path. TBH this is a change I would love to also make for the trainer (once we can without backwards compatibility).

Reply:
I like that idea, I can explore it, but would like to do that in a separate PR because it would change all tests. Since the implementation is copied from PL, I'd like to keep the changes in this initial PR as minimal as possible.